File: orchestrator/src/services/scoring_engine.py
"""5-dimensional message scoring engine powered by Claude with LLM safety features."""

import json
import logging
import asyncio
import time


from ..config.settings import settings
from .llm_provider import get_provider
from ..models.message import Message, ScoringContext, ScoringResult
from ..utils.prompt_sanitizer import sanitize_prompt, is_prompt_safe

logger = logging.getLogger(__name__)

# Scoring configuration
SCORING_TIMEOUT = 10  # seconds
SCORING_RETRY_ATTEMPTS = 3
SCORING_RETRY_DELAY = 1.0  # seconds

# Fallback score when LLM fails
FALLBACK_SCORE = 50


class ScoringEngine:
    """Evaluates candidate messages across 5 quality dimensions."""

    def __init__(self):
        """Initialize provider client via provider registry."""
        try:
            self.client = get_provider()
        except Exception as e:
            # If provider cannot be loaded, log and keep client as None -
            # the scoring functions will fallback gracefully.
            logger.error("LLM provider initialization failed: %s", e)
            self.client = None

        self.scoring_model = settings.SCORING_MODEL

    async def score_message(
        self,
        message: Message,
        context: ScoringContext,
    ) -> ScoringResult:
        """
        Score a single candidate message with safety features.

        Evaluation dimensions:
        - Relevance (35%): Directly addresses the room objective
        - Novelty (25%): Introduces new or useful information
        - Coherence (20%): Connects logically to prior discussion
        - Actionability (15%): Moves toward concrete outputs
        - Engagement (5%): Maintains viewer interest

        Safety Features:
        - Prompt injection prevention (sanitization)
        - Timeout protection (10s max)
        - Retry with exponential backoff
        - Fallback scoring (graceful degradation)
        - Request metrics logging

        Args:
            message: Candidate message to evaluate
            context: Room context and discussion history

        Returns:
            ScoringResult with breakdown and reasoning
        """
        start_time = time.time()
        
        # Layer 1: Sanitize user input to prevent prompt injection
        message_sanitization = sanitize_prompt(message.text)
        if not message_sanitization.is_safe:
            logger.warning(
                "Message contains safety violations",
                extra={
                    "message_id": message.id,
                    "violations": message_sanitization.violations,
                    "original_length": len(message.text),
                },
            )
        
        # Build prompt with sanitized message
        prompt = self._build_scoring_prompt(message, context)
        prompt_sanitization = sanitize_prompt(prompt)
        
        if not prompt_sanitization.is_safe:
            logger.warning(
                "Scoring prompt contains safety violations",
                extra={"violations": prompt_sanitization.violations},
            )

        # Layer 2: Score with timeout and retry logic
        score_data = await self._score_with_retry(prompt, message.id, context.room_id)
        
        duration_ms = (time.time() - start_time) * 1000

        # Log metrics
        logger.info(
            "Message scored",
            extra={
                "message_id": message.id,
                "agent_id": message.agent_id,
                "overall_score": score_data["overall_score"],
                "room_id": context.room_id,
                "duration_ms": int(duration_ms),
                "fallback_triggered": score_data.get("fallback_triggered", False),
            },
        )

        return ScoringResult(
            message_id=message.id,
            agent_id=message.agent_id,
            overall_score=score_data["overall_score"],
            relevance_score=score_data["relevance_score"],
            novelty_score=score_data["novelty_score"],
            coherence_score=score_data["coherence_score"],
            actionability_score=score_data["actionability_score"],
            engagement_score=score_data["engagement_score"],
            reasoning=score_data["reasoning"],
            strengths=score_data.get("strengths", []),
            weaknesses=score_data.get("weaknesses", []),
        )

    async def score_batch(
        self,
        messages: list[Message],
        context: ScoringContext,
    ) -> list[ScoringResult]:
        """
        Score multiple messages in parallel with concurrency control.
        Uses a semaphore to prevent overloading the LLM API.

        Args:
            messages: List of candidates
            context: Room context

        Returns:
            List of scoring results in same order as input
        """
        # Limit to 5 candidates per turn
        candidates = messages[: settings.MAX_CANDIDATES_PER_TURN]
        
        if not candidates:
            return []

        logger.info(
            "Starting parallel batch scoring",
            extra={
                "candidate_count": len(candidates),
                "room_id": context.room_id,
            },
        )

        # Control concurrency to respect rate limits (e.g., max 5 at once)
        semaphore = asyncio.Semaphore(5)

        async def _score_with_semaphore(msg):
            async with semaphore:
                return await self.score_message(msg, context)

        # Execute all scoring tasks in parallel
        tasks = [_score_with_semaphore(msg) for msg in candidates]
        results = await asyncio.gather(*tasks, return_exceptions=True)

        # Process results and handle potential exceptions
        final_results = []
        for i, res in enumerate(results):
            if isinstance(res, Exception):
                logger.error(
                    "Error scoring candidate in batch",
                    extra={
                        "message_id": candidates[i].id,
                        "error": str(res),
                    },
                )
                # Fallback for failed individual scoring
                final_results.append(self._create_fallback_result(candidates[i]))
            else:
                final_results.append(res)

        return final_results

    def _create_fallback_result(self, message: Message) -> ScoringResult:
        """Create a default fallback result for a failed scoring attempt."""
        return ScoringResult(
            message_id=message.id,
            agent_id=message.agent_id,
            overall_score=FALLBACK_SCORE,
            relevance_score=FALLBACK_SCORE,
            novelty_score=FALLBACK_SCORE,
            coherence_score=FALLBACK_SCORE,
            actionability_score=FALLBACK_SCORE,
            engagement_score=FALLBACK_SCORE,
            reasoning="Fallback score due to batch error",
            strengths=[],
            weaknesses=[],
        )

    def _build_scoring_prompt(self, message: Message, context: ScoringContext) -> str:
        """Build the scoring prompt for Claude."""
        transcript_str = "\n".join(
            [f"- {t['agent_id']}: {t['text']}" for t in context.transcript_history[-5:]]  # Last 5
        )

        prompt = f"""You are an evaluator for AI agent conversations. Score this message on 5 dimensions.

ROOM CONTEXT:
- Type: {context.room_type}
- Objective: {context.room_objective}

TRANSCRIPT (last 5 messages):
{transcript_str or "(empty)"}

CANDIDATE MESSAGE:
Agent: {message.agent_id}
Text: {message.text}

SCORING TASK:
Evaluate this message on these dimensions (0-100 each):

1. **Relevance (35% weight)**: Does it directly address the room objective?
2. **Novelty (25% weight)**: Does it introduce new or useful information?
3. **Coherence (20% weight)**: Does it connect logically to the discussion?
4. **Actionability (15% weight)**: Does it move toward concrete outputs?
5. **Engagement (5% weight)**: Does it maintain viewer interest?

RESPONSE FORMAT (JSON):
{{
  "relevance_score": <0-100>,
  "novelty_score": <0-100>,
  "coherence_score": <0-100>,
  "actionability_score": <0-100>,
  "engagement_score": <0-100>,
  "overall_score": <0-100>,
  "reasoning": "<brief explanation>",
  "strengths": ["<strength1>", "<strength2>"],
  "weaknesses": ["<weakness1>"]
}}

Respond ONLY with valid JSON."""

        return prompt

    async def _score_with_retry(
        self, prompt: str, message_id: str, room_id: str
    ) -> dict:
        """
        Score message with timeout, retry, and fallback logic.

        Strategy:
        1. Try to get LLM score with 10s timeout
        2. Retry up to 3 times with exponential backoff
        3. Return fallback score if all attempts fail
        4. Log all metrics for monitoring

        Args:
            prompt: Scoring prompt
            message_id: Message ID for logging
            room_id: Room ID for logging

        Returns:
            Score dictionary with overall_score, dimension scores, reasoning
        """
        last_error = None

        for attempt in range(SCORING_RETRY_ATTEMPTS):
            try:
                # Attempt scoring with timeout
                if self.client is None:
                    raise RuntimeError("LLM provider not configured")

                # Use provider's `messages_create` to allow multiple SDKs.
                # H7: Always include a system prompt to enforce JSON-only output;
                # without it some models return freeform text that fails JSON parsing.
                response = await asyncio.wait_for(
                    asyncio.to_thread(
                        self.client.messages_create,
                        model=self.scoring_model,
                        max_tokens=1024,
                        system=(
                            "You are a precise message scoring engine. "
                            "You MUST respond ONLY with valid JSON matching the specified schema. "
                            "Do not include any preamble, explanation, or markdown — just the JSON object."
                        ),
                        messages=[{"role": "user", "content": prompt}],
                    ),
                    timeout=SCORING_TIMEOUT,
                )

                # Parse response
                score_data = self._parse_scoring_response(response.content[0].text)
                return score_data

            except asyncio.TimeoutError:
                last_error = "timeout"
                logger.warning(
                    "Scoring timeout",
                    extra={
                        "message_id": message_id,
                        "room_id": room_id,
                        "attempt": attempt + 1,
                        "timeout_seconds": SCORING_TIMEOUT,
                    },
                )

                # Exponential backoff before retry
                if attempt < SCORING_RETRY_ATTEMPTS - 1:
                    await asyncio.sleep(SCORING_RETRY_DELAY * (2 ** attempt))

            except json.JSONDecodeError as e:
                last_error = "json_parse_error"
                logger.warning(
                    "Scoring response parse error",
                    extra={
                        "message_id": message_id,
                        "error": str(e),
                        "attempt": attempt + 1,
                    },
                )

                if attempt < SCORING_RETRY_ATTEMPTS - 1:
                    await asyncio.sleep(SCORING_RETRY_DELAY * (2 ** attempt))

            except Exception as e:
                last_error = str(e)
                logger.error(
                    "Scoring error",
                    extra={
                        "message_id": message_id,
                        "room_id": room_id,
                        "error": str(e),
                        "attempt": attempt + 1,
                    },
                )

                if attempt < SCORING_RETRY_ATTEMPTS - 1:
                    await asyncio.sleep(SCORING_RETRY_DELAY * (2 ** attempt))

        # All retries exhausted - return fallback score
        logger.error(
            "Scoring failed after all retries - using fallback",
            extra={
                "message_id": message_id,
                "room_id": room_id,
                "last_error": last_error,
                "fallback_score": FALLBACK_SCORE,
            },
        )

        return {
            "relevance_score": FALLBACK_SCORE,
            "novelty_score": FALLBACK_SCORE,
            "coherence_score": FALLBACK_SCORE,
            "actionability_score": FALLBACK_SCORE,
            "engagement_score": FALLBACK_SCORE,
            "overall_score": FALLBACK_SCORE,
            "reasoning": "Fallback score due to LLM unavailability",
            "strengths": [],
            "weaknesses": [],
            "fallback_triggered": True,
        }

    def _parse_scoring_response(self, response_text: str) -> dict:
        """
        Parse Claude's JSON scoring response.

        Handles:
        - Markdown code blocks (```json ... ```)
        - Invalid JSON (raises JSONDecodeError for retry)
        - Missing fields (use defaults)

        Args:
            response_text: Raw response from Claude

        Returns:
            Parsed score dictionary

        Raises:
            JSONDecodeError: If JSON parsing fails (triggers retry)
        """
        # Extract JSON from response (handle markdown code blocks)
        json_str = response_text.strip()
        if json_str.startswith("```"):
            json_str = json_str.split("```")[1]
            if json_str.startswith("json"):
                json_str = json_str[4:]
            json_str = json_str.strip()

        data = json.loads(json_str)  # Raises JSONDecodeError on invalid JSON

        # M5: Use .get() with safe defaults for all score fields so that a
        # partially-formed LLM response does not raise an unhandled KeyError.
        # Missing numeric fields fall back to FALLBACK_SCORE; missing text fields
        # fall back to sensible empty values. Log whenever fields are absent so
        # the LLM prompt or provider can be investigated.
        _missing: list[str] = []

        def _score(key: str) -> float:
            val = data.get(key)
            if val is None:
                _missing.append(key)
                return float(FALLBACK_SCORE)
            try:
                return max(0.0, min(100.0, float(val)))
            except (TypeError, ValueError):
                _missing.append(key)
                return float(FALLBACK_SCORE)

        relevance = _score("relevance_score")
        novelty = _score("novelty_score")
        coherence = _score("coherence_score")
        actionability = _score("actionability_score")
        engagement = _score("engagement_score")

        if _missing:
            logger.warning(
                "Scoring response missing expected fields — fallback values used",
                extra={"missing_fields": _missing},
            )

        # Compute weighted overall score
        overall = (
            relevance * settings.SCORING_WEIGHT_RELEVANCE
            + novelty * settings.SCORING_WEIGHT_NOVELTY
            + coherence * settings.SCORING_WEIGHT_COHERENCE
            + actionability * settings.SCORING_WEIGHT_ACTIONABILITY
            + engagement * settings.SCORING_WEIGHT_ENGAGEMENT
        )

        return {
            "relevance_score": relevance,
            "novelty_score": novelty,
            "coherence_score": coherence,
            "actionability_score": actionability,
            "engagement_score": engagement,
            "overall_score": overall,
            "reasoning": data.get("reasoning", "No reasoning provided"),
            "strengths": data.get("strengths", []),
            "weaknesses": data.get("weaknesses", []),
        }

File: backend/src/services/scoring/scoring-engine.ts
/**
 * Scoring Engine
 *
 * Port of orchestrator/src/services/scoring_engine.py
 *
 * Evaluates candidate messages across 5 quality dimensions using an LLM.
 * Produces a weighted overall score used by TurnSelector to pick the winner.
 *
 * Safety features:
 *  - Timeout + retry with exponential backoff
 *  - Fallback score (50) when LLM unavailable
 *  - Prompt injection mitigation (angle-bracket stripping)
 *  - JSON code-block unwrapping
 */

import { getLLMClient } from "./llm-provider.js";
import type { ScoringMessage, ScoringContext, ScoringResult, ScoringWeights } from "./types.js";
import { logger } from "../../utils/logger.js";

// ─── Config ───────────────────────────────────────────────────────────────────
// Model names are the only things that should vary between deployments.
// All tuning constants are encoded here — not in environment variables.

const SCORING_MODEL = process.env.SCORING_MODEL ?? "claude-3-5-sonnet-20241022";
const SCORING_TIMEOUT_MS = 10_000;   // 10 seconds per LLM call
const SCORING_RETRY_ATTEMPTS = 3;
const SCORING_RETRY_DELAY_MS = 1_000;
const FALLBACK_SCORE = 50;

// Default weights used when no room-type-specific weights are provided (e.g. in fallback paths).
const DEFAULT_WEIGHTS: ScoringWeights = {
  relevance:     0.35,
  novelty:       0.25,
  coherence:     0.20,
  actionability: 0.15,
  engagement:    0.05,
};

// ─── Prompt sanitizer (minimal injection prevention) ─────────────────────────

function sanitize(text: string): string {
  // Strip angle-bracket instruction-style injections: <anything>
  return text.replace(/<[^>]{0,200}>/g, "").trim();
}

// ─── ScoringEngine ────────────────────────────────────────────────────────────

export class ScoringEngine {
  /**
   * Score a single message with timeout, retry, and fallback.
   */
  async scoreMessage(
    message: ScoringMessage,
    context: ScoringContext,
  ): Promise<ScoringResult> {
    const start = Date.now();

    const weights = context.weights ?? DEFAULT_WEIGHTS;
    const prompt = this._buildScoringPrompt(message, context, weights);
    const scoreData = await this._scoreWithRetry(prompt, message.id, context.roomId, weights);

    logger.info("Message scored", {
      messageId: message.id,
      agentId: message.agentId,
      overallScore: scoreData.overall_score,
      roomId: context.roomId,
      durationMs: Date.now() - start,
      fallbackTriggered: scoreData.fallback_triggered ?? false,
    });

    return {
      messageId: message.id,
      agentId: message.agentId,
      overallScore: scoreData.overall_score,
      relevanceScore: scoreData.relevance_score,
      noveltyScore: scoreData.novelty_score,
      coherenceScore: scoreData.coherence_score,
      actionabilityScore: scoreData.actionability_score,
      engagementScore: scoreData.engagement_score,
      reasoning: scoreData.reasoning,
      strengths: scoreData.strengths ?? [],
      weaknesses: scoreData.weaknesses ?? [],
      isModerated: false,
      fallbackTriggered: scoreData.fallback_triggered ?? false,
    };
  }

  /**
   * Score multiple messages in parallel (max 5 concurrent, mirrors Python semaphore).
   */
  async scoreBatch(
    messages: ScoringMessage[],
    context: ScoringContext,
  ): Promise<ScoringResult[]> {
    const candidates = messages.slice(0, 10);

    if (candidates.length === 0) return [];

    logger.info("Starting parallel batch scoring", {
      candidateCount: candidates.length,
      roomId: context.roomId,
    });

    // Simple concurrency limiter (max 5 parallel LLM calls)
    const CONCURRENCY = 5;
    const results: ScoringResult[] = new Array(candidates.length);
    let index = 0;

    const worker = async () => {
      while (index < candidates.length) {
        const i = index++;
        try {
          results[i] = await this.scoreMessage(candidates[i], context);
        } catch (err) {
          logger.error("Error scoring candidate in batch", {
            messageId: candidates[i].id,
            error: err instanceof Error ? err.message : String(err),
          });
          results[i] = this._createFallbackResult(candidates[i]);
        }
      }
    };

    const workers = Array.from({ length: Math.min(CONCURRENCY, candidates.length) }, worker);
    await Promise.all(workers);

    return results;
  }

  // ─── Private helpers ─────────────────────────────────────────────────────────

  private _buildScoringPrompt(
    message: ScoringMessage,
    context: ScoringContext,
    weights: ScoringWeights,
  ): string {
    const transcriptStr = context.transcriptHistory
      .slice(-5)
      .map((t) => `- ${t.agentId}: ${t.text}`)
      .join("\n");

    const safeMsgText = sanitize(message.text);

    const pct = (w: number) => `${Math.round(w * 100)}%`;

    return `You are an evaluator for AI agent conversations. Score this message on 5 dimensions.

ROOM CONTEXT:
- Type: ${context.roomType}
- Objective: ${context.roomObjective}

TRANSCRIPT (last 5 messages):
${transcriptStr || "(empty)"}

CANDIDATE MESSAGE:
Agent: ${message.agentId}
Text: ${safeMsgText}

SCORING TASK:
Evaluate this message on these dimensions (0-100 each):

1. **Relevance (${pct(weights.relevance)} weight)**: Does it directly address the room objective?
2. **Novelty (${pct(weights.novelty)} weight)**: Does it introduce new or useful information?
3. **Coherence (${pct(weights.coherence)} weight)**: Does it connect logically to the discussion?
4. **Actionability (${pct(weights.actionability)} weight)**: Does it move toward concrete outputs?
5. **Engagement (${pct(weights.engagement)} weight)**: Does it maintain viewer interest?

RESPONSE FORMAT (JSON):
{
  "relevance_score": <0-100>,
  "novelty_score": <0-100>,
  "coherence_score": <0-100>,
  "actionability_score": <0-100>,
  "engagement_score": <0-100>,
  "overall_score": <0-100>,
  "reasoning": "<brief explanation>",
  "strengths": ["<strength1>", "<strength2>"],
  "weaknesses": ["<weakness1>"]
}

Respond ONLY with valid JSON.`;
  }

  private async _scoreWithRetry(
    prompt: string,
    messageId: string,
    roomId: string,
    weights: ScoringWeights,
  ): Promise<Record<string, unknown>> {
    let lastError: string = "";

    for (let attempt = 0; attempt < SCORING_RETRY_ATTEMPTS; attempt++) {
      try {
        const client = getLLMClient();

        const responseText = await Promise.race([
          client
            .messagesCreate({
              model: SCORING_MODEL,
              maxTokens: 1024,
              system:
                "You are a precise message scoring engine. " +
                "You MUST respond ONLY with valid JSON matching the specified schema. " +
                "Do not include any preamble, explanation, or markdown — just the JSON object.",
              messages: [{ role: "user", content: prompt }],
            })
            .then((r) => r.content[0]?.text ?? ""),
          new Promise<never>((_, reject) =>
            setTimeout(() => reject(new Error("scoring_timeout")), SCORING_TIMEOUT_MS),
          ),
        ]);

        return this._parseScoringResponse(responseText, messageId, weights);
      } catch (err) {
        const errMsg = err instanceof Error ? err.message : String(err);
        lastError = errMsg;

        if (errMsg === "scoring_timeout") {
          logger.warn("Scoring timeout", { messageId, roomId, attempt: attempt + 1, timeoutMs: SCORING_TIMEOUT_MS });
        } else {
          logger.error("Scoring error", { messageId, roomId, error: errMsg, attempt: attempt + 1 });
        }

        if (attempt < SCORING_RETRY_ATTEMPTS - 1) {
          await _sleep(SCORING_RETRY_DELAY_MS * Math.pow(2, attempt));
        }
      }
    }

    // All retries exhausted
    logger.error("Scoring failed after all retries — using fallback", {
      messageId,
      roomId,
      lastError,
      fallbackScore: FALLBACK_SCORE,
    });

    return _fallbackScoreData();
  }

  private _parseScoringResponse(
    responseText: string,
    messageId: string,
    weights: ScoringWeights,
  ): Record<string, unknown> {
    let json = responseText.trim();

    // Unwrap markdown code blocks: ```json ... ``` or ``` ... ```
    if (json.startsWith("```")) {
      const parts = json.split("```");
      json = parts[1] ?? "";
      if (json.startsWith("json")) json = json.slice(4);
      json = json.trim();
    }

    const data = JSON.parse(json) as Record<string, unknown>; // throws on invalid JSON → triggers retry

    const w = weights;
    const missing: string[] = [];

    const score = (key: string): number => {
      const val = data[key];
      if (val === undefined || val === null) {
        missing.push(key);
        return FALLBACK_SCORE;
      }
      const n = parseFloat(String(val));
      if (isNaN(n)) {
        missing.push(key);
        return FALLBACK_SCORE;
      }
      return Math.max(0, Math.min(100, n));
    };

    const relevance = score("relevance_score");
    const novelty = score("novelty_score");
    const coherence = score("coherence_score");
    const actionability = score("actionability_score");
    const engagement = score("engagement_score");

    if (missing.length > 0) {
      logger.warn("Scoring response missing expected fields — fallback values used", {
        messageId,
        missingFields: missing,
      });
    }

    const overall =
      relevance * w.relevance +
      novelty * w.novelty +
      coherence * w.coherence +
      actionability * w.actionability +
      engagement * w.engagement;

    return {
      relevance_score: relevance,
      novelty_score: novelty,
      coherence_score: coherence,
      actionability_score: actionability,
      engagement_score: engagement,
      overall_score: overall,
      reasoning: (data.reasoning as string) ?? "No reasoning provided",
      strengths: (data.strengths as string[]) ?? [],
      weaknesses: (data.weaknesses as string[]) ?? [],
    };
  }

  private _createFallbackResult(message: ScoringMessage): ScoringResult {
    return {
      messageId: message.id,
      agentId: message.agentId,
      overallScore: FALLBACK_SCORE,
      relevanceScore: FALLBACK_SCORE,
      noveltyScore: FALLBACK_SCORE,
      coherenceScore: FALLBACK_SCORE,
      actionabilityScore: FALLBACK_SCORE,
      engagementScore: FALLBACK_SCORE,
      reasoning: "Fallback score due to batch error",
      strengths: [],
      weaknesses: [],
      isModerated: false,
      fallbackTriggered: true,
    };
  }
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

function _sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function _fallbackScoreData(): Record<string, unknown> {
  return {
    relevance_score: FALLBACK_SCORE,
    novelty_score: FALLBACK_SCORE,
    coherence_score: FALLBACK_SCORE,
    actionability_score: FALLBACK_SCORE,
    engagement_score: FALLBACK_SCORE,
    overall_score: FALLBACK_SCORE,
    reasoning: "Fallback score due to LLM unavailability",
    strengths: [],
    weaknesses: [],
    fallback_triggered: true,
  };
}
